File: clients/ios/Lextures/Core/LMS/ProfileDepthModels.swift
import Foundation

// MARK: - Custom profile fields

struct ProfileFieldDefinition: Decodable, Identifiable, Hashable {
    var id: String
    var key: String
    var label: String
    var fieldType: String
    var selectOptions: [String]?
    var isRequired: Bool
}

struct ProfileFieldsResponse: Decodable {
    var fields: [ProfileFieldDefinition]
    var values: [String: JSONValue]
}

struct ProfileFieldsPatch: Encodable {
    var values: [String: JSONValue]
}

struct ProfileFieldsValuesResponse: Decodable {
    var values: [String: JSONValue]
}

// MARK: - Demographics (self-reported)

struct StudentDemographics: Codable, Equatable {
    var studentId: String?
    var freeLunch: Bool?
    var reducedLunch: Bool?
    var ellStatus: Bool?
    var disabilityStatus: Bool?
    var raceEthnicityCode: String?
    var homelessIndicator: Bool?
    var migrantIndicator: Bool?
    var dataSource: String?
    var updatedAt: String?
}

struct StudentDemographicsPatch: Encodable {
    var freeLunch: Bool?
    var reducedLunch: Bool?
    var ellStatus: Bool?
    var disabilityStatus: Bool?
    var raceEthnicityCode: String?
    var homelessIndicator: Bool?
    var migrantIndicator: Bool?
}

// MARK: - Research consent

enum ConsentDecision: String, Codable, CaseIterable {
    case granted, declined, withdrawn
}

struct ConsentStudy: Decodable, Identifiable, Hashable {
    var id: String
    var title: String
    var irbProtocol: String
    var consentText: String
    var dataUseDescription: String
    var status: String
}

struct ConsentStudiesResponse: Decodable {
    var studies: [ConsentStudy]
}

struct ConsentHistoryEntry: Decodable, Identifiable, Hashable {
    var id: String
    var studyId: String
    var studyTitle: String?
    var decision: ConsentDecision
    var createdAt: String

    var identifiableId: String { studyId }
}

struct ConsentHistoryResponse: Decodable {
    var history: [ConsentHistoryEntry]
}

struct ConsentRespondBody: Encodable {
    var decision: ConsentDecision
}

struct ConsentRecordResponse: Decodable {
    var record: ConsentRecord?
}

struct ConsentRecord: Decodable {
    var id: String
    var studyId: String
    var decision: ConsentDecision
}

/// Flexible JSON value for custom field payloads and review question payloads.
enum JSONValue: Codable, Hashable {
    case string(String)
    case bool(Bool)
    case number(Double)
    case object([String: JSONValue])
    case array([JSONValue])
    case null

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let boolValue = try? container.decode(Bool.self) {
            self = .bool(boolValue)
        } else if let numberValue = try? container.decode(Double.self) {
            self = .number(numberValue)
        } else if let stringValue = try? container.decode(String.self) {
            self = .string(stringValue)
        } else if let objectValue = try? container.decode([String: JSONValue].self) {
            self = .object(objectValue)
        } else if let arrayValue = try? container.decode([JSONValue].self) {
            self = .array(arrayValue)
        } else {
            self = .null
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let stringValue): try container.encode(stringValue)
        case .bool(let boolValue): try container.encode(boolValue)
        case .number(let numberValue): try container.encode(numberValue)
        case .object(let objectValue): try container.encode(objectValue)
        case .array(let arrayValue): try container.encode(arrayValue)
        case .null: try container.encodeNil()
        }
    }

    var isEmpty: Bool {
        switch self {
        case .null: return true
        case .string(let stringValue): return stringValue.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        case .array(let arrayValue): return arrayValue.isEmpty
        case .object(let objectValue): return objectValue.isEmpty
        case .bool, .number: return false
        }
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/ProfileDepthModels.kt
package com.lextures.android.core.lms

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull

@Serializable
data class ProfileFieldDefinition(
    val id: String,
    val key: String,
    val label: String,
    val fieldType: String,
    val selectOptions: List<String>? = null,
    val isRequired: Boolean = false,
)

@Serializable
data class ProfileFieldsResponse(
    val fields: List<ProfileFieldDefinition> = emptyList(),
    val values: Map<String, JsonElement> = emptyMap(),
)

@Serializable
data class ProfileFieldsPatch(
    val values: Map<String, JsonElement>,
)

@Serializable
data class ProfileFieldsValuesResponse(
    val values: Map<String, JsonElement> = emptyMap(),
)

@Serializable
data class StudentDemographics(
    val studentId: String? = null,
    val freeLunch: Boolean? = null,
    val reducedLunch: Boolean? = null,
    val ellStatus: Boolean? = null,
    val disabilityStatus: Boolean? = null,
    val raceEthnicityCode: String? = null,
    val homelessIndicator: Boolean? = null,
    val migrantIndicator: Boolean? = null,
    val dataSource: String? = null,
    val updatedAt: String? = null,
)

@Serializable
data class StudentDemographicsPatch(
    val freeLunch: Boolean? = null,
    val reducedLunch: Boolean? = null,
    val ellStatus: Boolean? = null,
    val disabilityStatus: Boolean? = null,
    val raceEthnicityCode: String? = null,
    val homelessIndicator: Boolean? = null,
    val migrantIndicator: Boolean? = null,
)

enum class ConsentDecision {
    @SerialName("granted") Granted,
    @SerialName("declined") Declined,
    @SerialName("withdrawn") Withdrawn,
}

@Serializable
data class ConsentStudy(
    val id: String,
    val title: String,
    val irbProtocol: String,
    val consentText: String,
    val dataUseDescription: String,
    val status: String,
)

@Serializable
data class ConsentStudiesResponse(
    val studies: List<ConsentStudy> = emptyList(),
)

@Serializable
data class ConsentHistoryEntry(
    val id: String,
    val studyId: String,
    val studyTitle: String? = null,
    val decision: ConsentDecision,
    val createdAt: String,
)

@Serializable
data class ConsentHistoryResponse(
    val history: List<ConsentHistoryEntry> = emptyList(),
)

@Serializable
data class ConsentRespondBody(
    val decision: ConsentDecision,
)

fun JsonElement?.displayText(fieldType: String): String? {
    val el = this ?: return null
    if (el is JsonNull) return null
    val prim = el as? JsonPrimitive ?: return null
    if (prim.isString) return prim.content
    prim.booleanOrNull?.let { return if (it) "true" else "false" }
    prim.doubleOrNull?.let { n ->
        return if (n % 1.0 == 0.0) n.toLong().toString() else n.toString()
    }
    return null
}